### PR TITLE
Add Adadelta Optimizer

### PR DESCRIPTION
**Pull request by dash2507**

### CODE REVIEW

File: Sources/TensorFlow/Optimizer.swift [hunk 1]
```diff
@@ -355,3 +355,91 @@ public class AdaGrad<Model: Layer>: Optimizer
...
+            accDeltaNew += (1 - Double(rho)) * (stepSize * stepSize)
+            accDelta[keyPath: kp] =  accDeltaNew
+        }
+    }
+
+    public func update(_ model: inout Model,
+                       along direction: Model.TangentVector) {
+        update(&model.allDifferentiableVariables, along: direction)
+    }
+}
```
Comment (Shashi456) [inline Sources/TensorFlow/Optimizer.swift]: Is this addition here by mistake?

File: Sources/TensorFlow/Optimizer.swift [hunk 2]
```diff
@@ -355,3 +355,91 @@ public class AdaGrad<Model: Layer>: Optimizer
         update(&model.allDifferentiableVariables, along: direction)
     }
 }
+
```
Comment (Shashi456) [inline Sources/TensorFlow/Optimizer.swift]: You need to add comments for the optimizer itself and a link to the paper that this is introduced in.

File: Sources/TensorFlow/Optimizer.swift [hunk 3]
```diff
@@ -355,3 +355,91 @@ public class AdaGrad<Model: Layer>: Optimizer
...
+    public var rho: Float
+    /// A small scalar added to the denominator to improve numerical stability.
+    public var epsilon: Float
+    /// The learning rate decay.
+    public var decay: Float
+    /// The current step.
+    public var step: Int = 0
+    public var accumulators: Model.AllDifferentiableVariables
+    public var accDelta: Model.AllDifferentiableVariables
+
```
Comment (rxwei) [inline Sources/TensorFlow/Optimizer.swift]: Remove extra line.

File: Sources/TensorFlow/Optimizer.swift [hunk 4]
```diff
@@ -355,3 +355,91 @@ public class AdaGrad<Model: Layer>: Optimizer
...
+    public var learningRate: Float
+    ///decay factor, corresponding to fraction of gradient to keep at each time step.
+    public var rho: Float
+    /// A small scalar added to the denominator to improve numerical stability.
+    public var epsilon: Float
+    /// The learning rate decay.
+    public var decay: Float
+    /// The current step.
+    public var step: Int = 0
+    public var accumulators: Model.AllDifferentiableVariables
```
Comment (rxwei) [inline Sources/TensorFlow/Optimizer.swift]: Could you add a doc comment that explains this property?

File: Sources/TensorFlow/Optimizer.swift [hunk 5]
```diff
@@ -355,3 +355,91 @@ public class AdaGrad<Model: Layer>: Optimizer
...
+    public var accDelta: Model.AllDifferentiableVariables
```
Comment (rxwei) [inline Sources/TensorFlow/Optimizer.swift]: - Could you add a doc comment that explains this property?
- Avoid abbreviations in this property name since `accDelta` is not a word of art and may be confusing.

File: Sources/TensorFlow/Optimizer.swift (hunk 1 above)
Comment (rxwei) [inline Sources/TensorFlow/Optimizer.swift]: Please leave a trailing empty line in the file.

File: Sources/TensorFlow/Optimizer.swift [hunk 6]
```diff
@@ -355,3 +355,102 @@ public class AdaGrad<Model: Layer>: Optimizer
...
+    public var learningRate: Float
+    ///decay factor, corresponding to fraction of gradient to keep at each time step.
+    public var rho: Float
+    /// A small scalar added to the denominator to improve numerical stability.
+    public var epsilon: Float
+    /// The learning rate decay.
+    public var decay: Float
+    /// The current step.
+    public var step: Int = 0
+    /// Accumulate Gradients
```
Comment (rxwei) [inline Sources/TensorFlow/Optimizer.swift]: Two suggestions on documentation comments:
- Properties should be described with a noun phrase ending with a period.
- Documentation comments for mathematical variables should use the term of art. In this case, we should switch to the accurate, proper terminology that's used by the paper.
```suggestion
    /// The accumulated, exponentially decaying average of squared gradients.
```

File: Sources/TensorFlow/Optimizer.swift [hunk 7]
```diff
@@ -355,3 +355,102 @@ public class AdaGrad<Model: Layer>: Optimizer
...
+    public var accumulators: Model.AllDifferentiableVariables
```
Comment (rxwei) [inline Sources/TensorFlow/Optimizer.swift]:
```suggestion
    public var averageSquared: Model.TangentVector
```

File: Sources/TensorFlow/Optimizer.swift [hunk 8]
```diff
@@ -355,3 +355,102 @@ public class AdaGrad<Model: Layer>: Optimizer
...
+    /// Accumulate Updates(here StepSizes)
+    public var updates: Model.AllDifferentiableVariables
```
Comment (rxwei) [inline Sources/TensorFlow/Optimizer.swift]:
```suggestion
    public var accumulatedDelta: Model.AllDifferentiableVariables
```

File: Sources/TensorFlow/Optimizer.swift (hunk 7 above)
Comment (rxwei) [reply Sources/TensorFlow/Optimizer.swift]: Although Keras uses the term `accumulator`, it is not very descriptive and does not connect with the paper very well. I'd suggest using the exact term mentioned in the paper:
> our methods implements this accumulation as an exponentially decaying **average of the squared gradients**.

This is why I suggest changing the name to `averageSquared`. Changing the type to `Model.TanegentVector` makes it more clearly convey that this is a **gradient** vector.

File: Sources/TensorFlow/Optimizer.swift [hunk 9]
```diff
@@ -355,3 +355,102 @@ public class AdaGrad<Model: Layer>: Optimizer
...
+    public var rho: Float
+    /// A small scalar added to the denominator to improve numerical stability.
+    public var epsilon: Float
+    /// The learning rate decay.
+    public var decay: Float
+    /// The current step.
+    public var step: Int = 0
+    /// Accumulate Gradients
+    public var accumulators: Model.AllDifferentiableVariables
+    /// Accumulate Updates(here StepSizes)
```
Comment (rxwei) [inline Sources/TensorFlow/Optimizer.swift]:
```suggestion
    /// The accumulated parameter updates.
```

File: Sources/TensorFlow/Optimizer.swift (hunk 8 above)
Comment (rxwei) [reply Sources/TensorFlow/Optimizer.swift]: Here using the word "delta" would make this variable better correspond to the `𝚫x` term in the paper.

File: Sources/TensorFlow/Optimizer.swift [hunk 10]
```diff
@@ -355,3 +355,102 @@ public class AdaGrad<Model: Layer>: Optimizer
...
+    public var updates: Model.AllDifferentiableVariables
+
+    public init(
+        for model: __shared Model,
+        learningRate: Float = 1,
+        rho: Float = 0.9,
```
Comment (rxwei) [inline Sources/TensorFlow/Optimizer.swift]: Keras uses `rho = 0.95` by default.

File: Sources/TensorFlow/Optimizer.swift [hunk 11]
```diff
@@ -355,3 +355,102 @@ public class AdaGrad<Model: Layer>: Optimizer
...
+
+    // TODO: Deprecate this when `Differentiable.AllDifferentiableVariables` is removed.
+    public func update(_ model: inout Model.AllDifferentiableVariables,
+                       along direction: Model.AllDifferentiableVariables) {
+        step += 1
+        let learningRate = self.learningRate * 1 / (1 + decay * Float(step))
+        
+        // Update Float & Double Tensor variables.
+        for kp in model.recursivelyAllWritableKeyPaths(to: Tensor<Float>.self) {
+            var accNew = rho * accumulators[keyPath: kp]
```
Comment (rxwei) [inline Sources/TensorFlow/Optimizer.swift]: For clarity, I'd suggest rewriting this step as

```swift
averageSquared[keyPath: kp] *= rho
averageSquared[keyPath: kp] += (1 - rho) * (direction[keyPath: kp] * direction[keyPath: kp])
```

This way, you won't need to define extra variables. Same suggestion for the `Tensor<Double>` handling block.

File: Sources/TensorFlow/Optimizer.swift [hunk 12]
```diff
@@ -355,3 +355,102 @@ public class AdaGrad<Model: Layer>: Optimizer
...
+            accNew += (1 - rho) * (direction[keyPath: kp] * direction[keyPath: kp])
+            accumulators[keyPath: kp] = accNew 
+
+            var stepSize =  direction[keyPath: kp] * sqrt(updates[keyPath: kp] + epsilon)
+            stepSize /=  sqrt(accumulators[keyPath: kp] + epsilon)
+            model[keyPath: kp] -= learningRate * stepSize
+
+            var updatesNew = updates[keyPath: kp] * rho
```
Comment (rxwei) [inline Sources/TensorFlow/Optimizer.swift]: I would suggest changing this step to the following:

```swift
accumulatedDelta *= rho
accumulatedDelta += (1 - rho) * stepSize.squared()
```
 Same suggestion for the `Tensor<Double>` handling block.

File: Sources/TensorFlow/Optimizer.swift [hunk 13]
```diff
@@ -355,3 +355,98 @@ public class AdaGrad<Model: Layer>: Optimizer
...
+            averageSquared[keyPath: kp].resetToZero()
+            accumulatedDelta[keyPath: kp].resetToZero()
+        }
+    }
+
+    // TODO: Deprecate this when `Differentiable.AllDifferentiableVariables` is removed.
+    public func update(_ model: inout Model.AllDifferentiableVariables,
+                       along direction: Model.AllDifferentiableVariables) {
+        step += 1
+        let learningRate = self.learningRate * 1 / (1 + decay * Float(step))
```
Comment (rxwei) [inline Sources/TensorFlow/Optimizer.swift]: `* 1` is redundant.
```suggestion
        let learningRate = self.learningRate / (1 + decay * Float(step))
```

File: Sources/TensorFlow/Optimizer.swift [hunk 14]
```diff
@@ -355,3 +355,98 @@ public class AdaGrad<Model: Layer>: Optimizer
...
+        
+        // Update Float & Double Tensor variables.
+        for kp in model.recursivelyAllWritableKeyPaths(to: Tensor<Float>.self) {
+            averageSquared[keyPath: kp] *= rho
+            averageSquared[keyPath: kp] +=
+                    (1 - rho) * (direction[keyPath: kp] * direction[keyPath: kp])
```
Comment (rxwei) [inline Sources/TensorFlow/Optimizer.swift]: Indent by 4.
```suggestion
                (1 - rho) * (direction[keyPath: kp] * direction[keyPath: kp])
```

File: Sources/TensorFlow/Optimizer.swift [hunk 15]
```diff
@@ -355,3 +355,98 @@ public class AdaGrad<Model: Layer>: Optimizer
...
+            var stepSize = direction[keyPath: kp] *
+                    sqrt(accumulatedDelta[keyPath: kp] + epsilon)
```
Comment (rxwei) [inline Sources/TensorFlow/Optimizer.swift]:
```suggestion
                sqrt(accumulatedDelta[keyPath: kp] + epsilon)
```

File: Sources/TensorFlow/Optimizer.swift [hunk 16]
```diff
@@ -355,3 +355,98 @@ public class AdaGrad<Model: Layer>: Optimizer
...
+            stepSize /= sqrt(averageSquared[keyPath: kp] + epsilon)
+            model[keyPath: kp] -= learningRate * stepSize
+            accumulatedDelta[keyPath: kp] *= rho
+            accumulatedDelta[keyPath: kp] += (1 - rho) * stepSize.squared()
+        }
+        for kp in model.recursivelyAllWritableKeyPaths(to: Tensor<Double>.self) {
+            averageSquared[keyPath: kp] *= Double(rho)
+            averageSquared[keyPath: kp] +=
+                    (1 - Double(rho)) * (direction[keyPath: kp] * direction[keyPath: kp])
```
Comment (rxwei) [inline Sources/TensorFlow/Optimizer.swift]:
```suggestion
                (1 - Double(rho)) * (direction[keyPath: kp] * direction[keyPath: kp])
```

File: Sources/TensorFlow/Optimizer.swift [hunk 17]
```diff
@@ -355,3 +355,98 @@ public class AdaGrad<Model: Layer>: Optimizer
...
+            var stepSize = direction[keyPath: kp] *
+                    sqrt(accumulatedDelta[keyPath: kp] + Double(epsilon))
```
Comment (rxwei) [inline Sources/TensorFlow/Optimizer.swift]:
```suggestion
                sqrt(accumulatedDelta[keyPath: kp] + Double(epsilon))
```

File: Sources/TensorFlow/Optimizer.swift [hunk 18]
```diff
@@ -355,3 +355,98 @@ public class AdaGrad<Model: Layer>: Optimizer
         update(&model.allDifferentiableVariables, along: direction)
     }
 }
+
+/// Adadelta Optimizer
```
Comment (rxwei) [inline Sources/TensorFlow/Optimizer.swift]: - Use the canonical name (the name in the paper) to describe this class. 
- End all comments with a period.
- Make the "O" in "Optimizer" be lowercased, for consistency with API documentation on other optimizers in this file.
```suggestion
/// ADADELTA optimizer.
```

File: Sources/TensorFlow/Optimizer.swift [hunk 19]
```diff
@@ -355,3 +355,98 @@ public class AdaGrad<Model: Layer>: Optimizer
...
+///
+/// It is a method that uses the magnitude of recent gradients
```
Comment (rxwei) [inline Sources/TensorFlow/Optimizer.swift]: API documentation should fit to size (i.e. to 100 columns).

I'd suggest changing the summary to the following, similar to the Keras documentation.
> ADADELTA is a more robust extension of AdaGrad that adapts learning rates based on a moving window of gradient updates, instead of accumulating all past gradients. This way, Adadelta continues learning even when many updates have been done. Compared to AdaGrad, in the original version of ADADELTA you don't have to set an initial learning rate. In this version, initial learning rate and decay factor can be set, as in most other optimizers.

File: Sources/TensorFlow/Optimizer.swift [hunk 20]
```diff
@@ -355,3 +355,98 @@ public class AdaGrad<Model: Layer>: Optimizer
...
+        }
+    }
+
+    // TODO: Deprecate this when `Differentiable.AllDifferentiableVariables` is removed.
+    public func update(_ model: inout Model.AllDifferentiableVariables,
+                       along direction: Model.AllDifferentiableVariables) {
+        step += 1
+        let learningRate = self.learningRate * 1 / (1 + decay * Float(step))
+        
+        // Update Float & Double Tensor variables.
```
Comment (rxwei) [inline Sources/TensorFlow/Optimizer.swift]:
```suggestion
        // Update `Tensor<Float>` and `Tensor<Double>` variables.
```

File: Sources/TensorFlow/Optimizer.swift [hunk 21]
```diff
@@ -355,3 +355,100 @@ public class AdaGrad<Model: Layer>: Optimizer
         update(&model.allDifferentiableVariables, along: direction)
     }
 }
+
+/// ADADELTA optimizer.
+///
+/// ADADELTA is a more robust extension of AdaGrad that adapts learning rates
```
Comment (rxwei) [inline Sources/TensorFlow/Optimizer.swift]: This block of API documentation has lines whose length varies significantly from lines to lines. Could you make sure  each line is filled with words until it reaches 100 columns? This is what I meant by suggesting "fit to size". Hope this clarifies things!

**Comment by rxwei** [comment]
Thanks @dan-zheng for reordering the class definitions!

**Comment by Shashi456** [comment]
#127
